File: generate-cache-dot.js
const fs = require('fs');
const path = require('path');

// Read the index.json file
const indexPath = '/c/rnd/claude/mcp/ppom_mcp/plugins/php-index-generator/output/index.json';
const indexData = JSON.parse(fs.readFileSync(indexPath, 'utf8'));

// Extract cache-related functions
const cacheRegex = /cache/i;
const cacheFunctions = {};
const cacheFiles = {};

// Find all symbols matching cache pattern
Object.entries(indexData.symbols).forEach(([key, value]) => {
  if (cacheRegex.test(value.name)) {
    cacheFunctions[key] = value;
  }
});

// Find all files matching cache pattern
Object.entries(indexData.files).forEach(([key, value]) => {
  if (cacheRegex.test(key)) {
    cacheFiles[key] = value;
  }
});

// Extract dependencies from functionCalls section
const dependencies = indexData.dependencies || {};
const cacheDepMap = {};

Object.entries(dependencies).forEach(([filePath, fileData]) => {
  if (cacheRegex.test(filePath)) {
    fileData.functionCalls.forEach(call => {
      if (!cacheDepMap[filePath]) {
        cacheDepMap[filePath] = [];
      }
      cacheDepMap[filePath].push({
        name: call.name,
        type: call.type,
        builtin: call.builtin,
        source: call.source
      });
    });
  }
});

// Generate DOT format
let dotContent = 'digraph CacheDependencies {\n';
dotContent += '  rankdir=LR;\n';
dotContent += '  node [shape=box, style=rounded];\n';
dotContent += '  edge [color=gray];\n\n';

// Add nodes for cache functions
const nodeMap = {};
let nodeId = 0;

Object.entries(cacheFunctions).forEach(([key, func]) => {
  const nodeName = `node${nodeId}`;
  nodeMap[func.name] = nodeName;

  const file = path.basename(func.file || '');
  const label = `${func.name}\\n(${file})`;
  dotContent += `  ${nodeName} [label="${label}", fillcolor="#E3F2FD", style="filled,rounded"];\n`;
  nodeId++;
});

dotContent += '\n';

// Add edges for dependencies
Object.entries(cacheDepMap).forEach(([filePath, calls]) => {
  calls.forEach(call => {
    if (nodeMap[call.name] && !call.builtin) {
      // Find which cache function calls this
      Object.entries(cacheFunctions).forEach(([key, func]) => {
        const funcFile = func.file.replace(/\\/g, '/');
        const depFile = filePath.replace(/\\/g, '/');

        if (funcFile.includes(depFile.split('/').pop())) {
          const sourceNode = nodeMap[func.name];
          const targetNode = nodeMap[call.name];

          if (sourceNode && targetNode) {
            dotContent += `  ${sourceNode} -> ${targetNode} [label="${call.type}", color="${call.builtin ? '#CCC' : '#666'}"];\n`;
          }
        }
      });
    }
  });
});

dotContent += '}\n';

// Create output directory
const outputDir = '/c/Users/aired/AppData/Roaming/Claude/local-agent-mode-sessions/skills-plugin/03887657-ce9f-4a37-a469-3b10e0effb35/2e999d38-475d-40e8-8a21-5bfea2929df8/php-dependency-mapper-workspace/iteration-1/eval-dot-export/without_skill/outputs/';

if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

// Write DOT file
const outputFile = path.join(outputDir, 'cache_dependencies.dot');
fs.writeFileSync(outputFile, dotContent, 'utf8');

console.log(`DOT file generated: ${outputFile}`);
console.log(`Cache functions found: ${Object.keys(cacheFunctions).length}`);
console.log(`Cache files found: ${Object.keys(cacheFiles).length}`);

// Also write a summary
const summary = {
  total_cache_functions: Object.keys(cacheFunctions).length,
  total_cache_files: Object.keys(cacheFiles).length,
  functions: Object.values(cacheFunctions).map(f => ({
    name: f.name,
    file: f.file,
    line: f.line,
    type: f.type
  })),
  files: Object.keys(cacheFiles)
};

const summaryFile = path.join(outputDir, 'cache_analysis_summary.json');
fs.writeFileSync(summaryFile, JSON.stringify(summary, null, 2), 'utf8');

console.log(`Summary file generated: ${summaryFile}`);

File: generate_cache_dot.py
#!/usr/bin/env python3
"""
PHP Cache Dependency Graph Generator
Generates DOT format dependency graph for cache-related functions
"""

import re
import os
from pathlib import Path
from collections import defaultdict

class PHPDependencyAnalyzer:
    def __init__(self, source_dir):
        self.source_dir = source_dir
        self.functions = {}
        self.calls = defaultdict(set)
        self.cache_functions = set()

    def find_php_files(self, pattern='cache'):
        """Find PHP files containing cache-related code"""
        php_files = []
        for root, dirs, files in os.walk(self.source_dir):
            for file in files:
                if file.endswith('.php'):
                    filepath = os.path.join(root, file)
                    if pattern.lower() in file.lower():
                        php_files.append(filepath)
        return php_files

    def extract_functions(self, content):
        """Extract function definitions from PHP content"""
        functions = {}

        # Pattern for function declarations
        func_pattern = r'^\s*(?:public\s+|private\s+|protected\s+)?function\s+(\w*cache\w*)\s*\('

        for i, line in enumerate(content.split('\n'), 1):
            match = re.match(func_pattern, line, re.IGNORECASE)
            if match:
                func_name = match.group(1)
                functions[func_name] = i
                self.cache_functions.add(func_name)

        return functions

    def extract_calls(self, content, source_function):
        """Extract function calls from PHP content"""
        calls = set()

        # Patterns for function calls
        patterns = [
            r'\$this->(\w*cache\w*)\s*\(',  # Object method calls
            r'(\w*cache\w*)\s*\(',           # Direct function calls
        ]

        for pattern in patterns:
            matches = re.finditer(pattern, content, re.IGNORECASE)
            for match in matches:
                called_func = match.group(1)
                if called_func.lower() != source_function.lower():
                    calls.add(called_func)

        return calls

    def analyze_files(self):
        """Analyze all cache-related PHP files"""
        php_files = self.find_php_files('cache')

        for filepath in php_files:
            try:
                with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
                    content = f.read()

                # Extract class name for context
                class_match = re.search(r'class\s+(\w*[Cc]ache\w*)', content)
                class_name = class_match.group(1) if class_match else None

                # Extract functions
                functions = self.extract_functions(content)

                for func_name, line_num in functions.items():
                    full_name = f"{class_name}::{func_name}" if class_name else func_name
                    self.functions[full_name] = {
                        'file': filepath,
                        'line': line_num,
                        'class': class_name
                    }

                    # Extract calls from the function
                    calls = self.extract_calls(content, func_name)
                    for called_func in calls:
                        self.calls[full_name].add(called_func)

            except Exception as e:
                print(f"Error processing {filepath}: {e}")

    def generate_dot(self):
        """Generate DOT format graph"""
        lines = [
            'digraph "PHP Cache Dependencies" {',
            '  rankdir=LR;',
            '  node [shape=box, style=rounded];',
            '  edge [color=darkblue];',
            '',
            '  // Cache Functions',
        ]

        # Add all cache functions as nodes
        for func_name in sorted(self.cache_functions):
            safe_name = self._sanitize_name(func_name)
            lines.append(f'  "{safe_name}" [label="{func_name}", color=lightblue, style="rounded,filled"];')

        lines.append('')
        lines.append('  // Function Calls')

        # Add all function calls as edges
        added_edges = set()
        for caller, callees in sorted(self.calls.items()):
            caller_safe = self._sanitize_name(caller)
            for callee in sorted(callees):
                callee_safe = self._sanitize_name(callee)
                edge = f'  "{caller_safe}" -> "{callee_safe}";'
                if edge not in added_edges:
                    lines.append(edge)
                    added_edges.add(edge)

        lines.append('')
        lines.append('  // Legend')
        lines.append('  subgraph cluster_legend {')
        lines.append('    label = "Legend";')
        lines.append('    style = "rounded,dashed";')
        lines.append('    "function_node" [label="Function", color=lightblue, style="rounded,filled"];')
        lines.append('  }')
        lines.append('}')

        return '\n'.join(lines)

    def _sanitize_name(self, name):
        """Sanitize names for DOT format"""
        # Replace special characters
        return re.sub(r'[^a-zA-Z0-9_:]', '_', name)


def main():
    source_dir = r'C:\rnd\claude\mcp\ppom_mcp\work'
    output_dir = r'C:\Users\aired\AppData\Roaming\Claude\local-agent-mode-sessions\skills-plugin\03887657-ce9f-4a37-a469-3b10e0effb35\2e999d38-475d-40e8-8a21-5bfea2929df8\php-dependency-mapper-workspace\iteration-1\eval-dot-export\with_skill\outputs'

    # Create output directory
    os.makedirs(output_dir, exist_ok=True)

    # Analyze dependencies
    analyzer = PHPDependencyAnalyzer(source_dir)
    analyzer.analyze_files()

    # Generate DOT output
    dot_content = analyzer.generate_dot()

    # Save to file
    output_file = os.path.join(output_dir, 'cache_dependencies.dot')
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(dot_content)

    print(f"DOT file generated: {output_file}")
    print(f"Total cache functions found: {len(analyzer.cache_functions)}")
    print(f"Total function calls: {sum(len(callees) for callees in analyzer.calls.values())}")

    # Display summary
    print("\n=== Cache Functions ===")
    for func in sorted(analyzer.cache_functions):
        info = analyzer.functions.get(func, {})
        print(f"  {func}")
        if 'file' in info:
            print(f"    File: {info['file']}")
            print(f"    Line: {info['line']}")


if __name__ == '__main__':
    main()
